coreclient/src/clients/message.rs

```diff
@@ -308,27 +374,76 @@ struct SentMessage {
 }
 
 impl SentMessage {
-    async fn mark_as_sent_and_read(
+    async fn mark_as_sent(
         self,
-        connection: &mut sqlx::SqliteConnection,
+        txn: &mut SqliteTransaction<'_>,
         notifier: &mut StoreNotifier,
+        own_user: &UserId,
     ) -> anyhow::Result<ConversationMessage> {
         let Self {
             mut conversation_message,
             ds_timestamp,
         } = self;
 
         conversation_message
-            .mark_as_sent(connection, notifier, ds_timestamp)
+            .mark_as_sent(&mut *txn, notifier, ds_timestamp)
             .await?;
+
+        // Note: even though the message was already marked as read, we still need to move the last
+        // read timestamp down. After a message was sent to DS, it is marked as read, which updates
+        // its timestamp to the timestamp returned by DS.
         Conversation::mark_as_read_until_message_id(
-            connection,
+            txn,
             notifier,
             conversation_message.conversation_id(),
             conversation_message.id(),
+            own_user,
         )
         .await?;
 
         Ok(conversation_message)
     }
 }
+
+/// Not yet sent receipt message consisting of the content to send and a local message status
+/// report.
+struct UnsentReceipt {
+    report: MessageStatusReport,
+    content: MimiContent,
+}
+
+impl UnsentReceipt {
+    fn new<'a>(
+        statuses: impl IntoIterator<Item = (&'a MimiId, MessageStatus)>,
+    ) -> anyhow::Result<Option<Self>> {
+        let report = MessageStatusReport {
+            statuses: statuses
+                .into_iter()
+                .map(|(id, status)| PerMessageStatus {
+                    mimi_id: id.as_ref().to_vec().into(),
+                    status,
+                })
+                .collect(),
+        };
+
+        if report.statuses.is_empty() {
+            return Ok(None);
+        }
+
+        let salt: [u8; 16] = RustCrypto::default().random_array()?;
```

Comment:
I don't know what RustCrypto does, but we should probably use the same methods for generating salts everywhere. I used phnxcommon::crypto::secrets::Secret::<16>::random()?.secret(),